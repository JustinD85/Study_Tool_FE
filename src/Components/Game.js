import React, { Component } from 'react';
import Fish from './Fish';
import uuid from 'uuid'

class Game extends Component {
  constructor(props) {
    super(props);
    this.state = {
      intervalId: null,
      wordWorth: 100,
      possibleWords: [],
      possibleDefinitions: [],
      fishWords: [],
      currentDefinition: null,
      currentWord: null,
      currentTopic: null
    }
  }

  randNumGen(num) {
    return Math.floor(Math.random() * num);
  }

  selectTopic = () => {
    const { topics, user } = this.props.info;
    let allTopics = [...topics.keys()];
    let index = Math.floor(Math.random() * allTopics.length);
    let currentTopic = allTopics[index];
    let { possibleDefinitions, possibleWords } = this.state;

    topics.get(currentTopic).forEach((section) => {
      section.get('vocabulary').forEach((definition, word) => {
        if (!user[currentTopic].words.includes(word)) {
          possibleWords.push(word);
          const isCSSVocab = typeof definition === 'object';
          if (isCSSVocab) {
            definition = definition.get('description');
          }
          possibleDefinitions.push(definition)
        };
      });
    });

    this.setState({
      currentTopic, possibleDefinitions, possibleWords
    }, this.setFishWords);
  }

  setFishWords() {
    let { possibleDefinitions, possibleWords } = this.state;
    let index = this.randNumGen(possibleWords.length);
    let currentWord = possibleWords.splice(index, 1).toString();
    let currentDefinition = possibleDefinitions.splice(index, 1).toString();
    let fishWords = [];
    //sets fish words
    for (let i = 0; i < 3; i++) {
      index = this.randNumGen(possibleWords.length);
      fishWords.push(...possibleWords.splice(index, 1));
      possibleDefinitions.splice(index, 1);
    };

    fishWords.push(currentWord);

    //randomizer
    for (let i = 0; i < 8; i++) {
      fishWords.unshift(...fishWords.splice(this.randNumGen(fishWords.length), 1));
    }

    this.setState({
      currentWord,
      currentDefinition,
      fishWords
    });
  }

  decrementPoints = () => {
    const wordWorth = this.state.wordWorth - 1;
    if (wordWorth < 1) {
      this.setUserScore();
    } else {
      this.setState({ wordWorth });
    }
  }

  componentDidMount() {
    this.selectTopic();
    const intervalId = setInterval(this.decrementPoints, 1000);
    this.setState({ intervalId });
  }

  componentWillUnmount() {
    clearInterval(this.state.intervalId);
  }

  setUserScore = (e) => {
    let { currentWord, currentTopic, wordWorth } = this.state;
    let { user, handleSaveToStorage } = this.props.info;
    let userWords = user[currentTopic].words;
    let selWord = e ? e.target.closest('.Fish').dataset.id : '';

    if (selWord === currentWord) {
      user.score += wordWorth;
      userWords.push(currentWord);
    } else {
      if (userWords.includes(currentWord)) {
        userWords.splice(userWords.findIndex(currentWord), 1);
      }
    }

    this.selectTopic();
    handleSaveToStorage();
    clearInterval(this.state.intervalId);
    const intervalId = setInterval(this.decrementPoints, 1000);
    this.setState({ intervalId, wordWorth:100 })
  }


  render() {
    const { score } = this.props.info.user;
    const { wordWorth, currentDefinition, fishWords } = this.state;
    return (
      <div className="Game">
        <div className="game-sky">
          <span className="game-definition game-status"> {currentDefinition}</span>
          <span className="game-definition-score game-status">pointValue:{wordWorth}</span>
          <span className="game-user-score game-status">Score:{score}</span>
        </div>
        <div className="game-ocean">
          {fishWords.map((word, i) => <Fish key={uuid()}
            word={word} index={i}
            top={'170px'}
            setUserScore={this.setUserScore}

          />)}
        </div>
      </div>
    );
  }
}

export default Game;