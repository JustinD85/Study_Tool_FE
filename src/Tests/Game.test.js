import React from 'react';
import Game from '../Components/Game';
import { shallow } from 'enzyme';

describe('<Game/>', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Game />
    );
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});