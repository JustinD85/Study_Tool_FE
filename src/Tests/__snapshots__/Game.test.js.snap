// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game/> should match the snapshot 1`] = `
<div
  className="Game"
>
  <div
    className="game-sky"
  >
    <span
      className="game-definition game-status"
    >
       
      Assigning a value to a variable
    </span>
    <span
      className="game-definition-score game-status"
    >
      pointValue:
      100
    </span>
    <span
      className="game-user-score game-status"
    >
      Score:
      150
    </span>
  </div>
  <div
    className="game-ocean"
  >
    <Fish
      index={0}
      key="946114f9-be0a-4258-bb3c-3d5f9c1a0c98"
      setUserScore={[Function]}
      top="170px"
      word="hoisting"
    />
    <Fish
      index={1}
      key="7e4dbd40-2368-48be-86da-8058a2bfe433"
      setUserScore={[Function]}
      top="170px"
      word="variable"
    />
    <Fish
      index={2}
      key="47352ca0-2cfa-4220-a948-194976151cfb"
      setUserScore={[Function]}
      top="170px"
      word="function"
    />
    <Fish
      index={3}
      key="fb9dfb27-bf29-4320-88a9-17c678ea997a"
      setUserScore={[Function]}
      top="170px"
      word="assignment"
    />
  </div>
</div>
`;
